### PR TITLE
fix(chat): route streamed-chat parser through safe_index + surface "er" frames

### PR DESCRIPTION
**Pull request by teng-lin**
## Summary

The streamed-chat parser in `_chat_protocol.py` (the highest-traffic decode path) hand-rolled positional indexing and did **not** use `safe_index`, so it silently bypassed `NOTEBOOKLM_STRICT_DECODE`. On a Google wire-shape change, a populated `wrb.fr` answer record whose answer row was no longer a list collapsed to an **empty answer** instead of raising drift. It also only inspected `wrb.fr` frames and silently skipped `er` error frames, so a server-side chat error degraded into a generic "empty response" / parse failure.

This is the v0.6.0 architecture gap-review remediation for the **Tier-1 / existential** finding: `_chat_protocol` is explicitly named as unmigrated strict-decode debt in **ADR-011:38** (`docs/adr/0011-schema-validation-policy.md`), alongside `_artifact_downloads` / `_artifact_polling`.

## What changed

**(a) Route the load-bearing answer descents through `safe_index`** (frame tag, answer row, answer-text leaf), threading a `source` label — mirroring how `rpc/decoder.py` consumes `safe_index`.
- The existing `isinstance`/`len` guards already pre-empt out-of-bounds reads, so the **happy path is byte-for-byte unchanged**.
- A *populated* `wrb.fr` record whose answer row is **not a list** is now treated as genuine Google-side drift and raises `UnknownRPCMethodError` under the default strict-decode mode. `safe_index` cannot enforce the list *type* (a `str` answer row is still indexable at `[0]`), so the contract is checked explicitly with the same `is_strict_decode_enabled()` gate `safe_index` uses.
- Soft-mode (`NOTEBOOKLM_STRICT_DECODE=0`) preserves the legacy skip-and-degrade behavior.
- Empty (`[]`) answer rows stay tolerated as degenerate/heartbeat records.

**(b) Detect `er` frames** and surface the embedded error/code as a `ChatError` instead of silently skipping them. The optional code slot is read with a length guard — its absence is not drift, since the `er` frame itself is the error signal.

Per-citation descents (`parse_citations`, `parse_single_citation`, `extract_text_passages`, `collect_texts_from_nested`) keep their tolerant degrade-without-raising contract, which is pinned by existing tests; they are deliberately left for a follow-up (the gap-review explicitly allows migrating the load-bearing answer + citation entry points first).

## Tests

- All existing `tests/unit/test_streaming_chat_protocol.py` cases stay green (citations / scores / ranges / XSSI / empty-raise / `UserDisplayableError`).
- Added: drifted chat envelope raises `UnknownRPCMethodError` under strict-decode (parametrized over scalar / dict / int / null answer rows); soft-mode degrades to empty; `[[]]` heartbeat tolerated; an `er` chat frame surfaces a `ChatError` with the embedded code (and a short `er` frame with no code still raises `ChatError`).
- Updated the one characterization test in `test_chat_characterization.py` that pinned the **old silent-skip** behavior to assert the new strict-raises / soft-skips contract (it characterized exactly the bug being fixed).

## Gates

- `ruff check .` + `ruff format --check .` — pass
- `mypy src/notebooklm --ignore-missing-imports` — pass
- `pytest tests/unit/test_streaming_chat_protocol.py tests/unit/test_chat.py tests/unit/test_chat_characterization.py` — 133 passed

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Release Notes

* **Bug Fixes**
  * Chat responses now explicitly surface server-side RPC errors instead of silently failing with empty results.
  * Improved handling of malformed server data with configurable strict and lenient parsing modes.

* **Tests**
  * Added test coverage for error handling and edge-case behavior in chat streaming.

[Review Change Stack](https://app.coderabbit.ai/change-stack/teng-lin/notebooklm-py/pull/1219?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by teng-lin** [comment]
@claude review

File: src/notebooklm/_chat_protocol.py [hunk 1]
```diff
@@ -344,6 +392,29 @@ def _extract_chunk_with_parseable(
...
+def _raise_chat_error_frame(item: list) -> None:
+    """Surface a server-side ``"er"`` error frame as a ``ChatError``.
+
+    The streamed batchexecute backend emits ``["er", rpc_id, code, ...]``
+    frames when the RPC itself failed. The previous parser only inspected
+    ``"wrb.fr"`` frames and silently skipped these, so a real server-side
+    chat error collapsed into the generic ``ChatResponseParseError`` (or an
+    empty answer). Threading the ``code`` slot through ``safe_index`` keeps
+    the read drift-aware; the embedded code/message is included verbatim so
+    callers see the actual failure instead of a generic parse error.
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_chat_protocol.py]: Addressed in commit 4fea3dbe: the docstring now matches the implementation. Rewrote it to state that the optional `code` slot is read with an explicit length guard rather than `safe_index` (an absent code on a short "er" frame is normal and must not be treated as schema drift, since the frame is itself the error signal). Docstring-only change; no behavior impact.

**Comment by teng-lin** [comment]
@claude review

**Comment by teng-lin** [comment]
Addressed both findings from the latest review in commit 6d028d00:

- **Finding 2** (`-> NoReturn`): `_raise_chat_error_frame` is now annotated `-> NoReturn` (with `NoReturn` added to the typing import), so mypy treats the call as a flow-control narrowing point. mypy is green.
- **Finding 3** (re-traversal nit): the answer-text leaf now reads from the already-held `first` (`safe_index(first, 0)`) instead of re-descending `inner_data[0][0]`. The `len(first) > 0` guard makes the two equivalent, so the extracted answer text is byte-for-byte identical on the happy path; drift diagnostics now localize to `(0,)`.

Both are behavior-preserving on the chat happy path — all 443 chat/stream/citation/protocol tests stay green, and ruff + mypy pass.

@claude review

**Comment by teng-lin** [comment]
This PR was squash-merged as e987bbfb. The gemini docstring finding (4fea3dbe) made it into the merge snapshot. The two remaining claude-review findings (Finding 2: `-> NoReturn`; Finding 3: reuse the held answer row) landed on the branch a few seconds *after* the merge snapshot, so they are not yet in main. Both are non-functional and behavior-preserving on the chat happy path; tracked for re-application in #1230.
